Выполнил практическое задание 14-го спринта.

MaximusDenisov-YP opened the pull request:
(Реализовал 1-ю часть ShareIt)

- Реализовал MVC в рамках структуры DDD
- Подготовил приложение ко 2-й части практического задания
- Из нового применил разделение DTO и Entity и добавил соответствующие мапперы

MaximusDenisov-YP committed:
Выполнил практическое задание 13-го спринта.
(Реализовал 1-ю часть ShareIt)

- Реализовал MVC в рамках структуры DDD
- Подготовил приложение ко 2-й части практического задания
- Из нового применил разделение DTO и Entity и добавил соответствующие мапперы

MaximusDenisov-YP committed:
Удалил пустые классы созданные вместе с проектом

avfyodorov commented on src/main/java/ru/practicum/shareit/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,71 @@
+                        exc.getMessage()),
+                HttpStatus.CONFLICT
+        );
+    }
```
Так же можно добавить хендлер для обработки всех необработанных исключений, ловить  Exception, а возвращать статус 500 INTERNAL_SERVER_ERROR

avfyodorov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,49 @@
+    }
+
+    @GetMapping("/{itemId}")
+    public Item getItemById(@PathVariable Long itemId, @RequestHeader("X-Sharer-User-Id") Long userId) {
```
Item 

В этом методе выходной параметр должен  иметь тип ItemDto. Вызывается он откуда-то снаружи,  из фронта и передаётся наружу. А фронт,в свою очередь, работает с объектами типа дто, например,ItemDto и не имеет доступа  к родным объектам базы данных (Item). По крайней мере, не должен иметь.

avfyodorov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,49 @@
+    }
+
+    @GetMapping("/search")
+    public List<Item> getSearchingItemsByText(@RequestParam String text) {
```
public List<Item> --> ItemDto

Посмотрите, пожалуйста,  аналогичные ситуации и в других методах и в других контроллерах.

avfyodorov commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -0,0 +1,24 @@
+import ru.practicum.shareit.item.dto.ItemDto;
+import ru.practicum.shareit.item.entity.Item;
+
+public class ItemMapper {
```
В принципе, для мапперов можно добавить аннотацию: 

@NoArgsConstructor(access = AccessLevel.PRIVATE) или @UtilityClass

Поскольку все методы Статик, то нет никакого смысла разрешать  создавать экземпляры этого класса. 
 необязательно,  на Ваше усмотрение.

avfyodorov commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Override
+    public List<Item> getSearchingItemsByText(String text) {
+        return items.values().stream()
```
text.toLowerCase()

Можно было бы преобразовать к нижнему регистру один раз до начала цикла, А так получается каждый раз по два раза :)

avfyodorov reviewed (commented):
Добрый вечер, Максим! 

Хорошая работа, всего несколько небольших уточнений.

MaximusDenisov-YP committed:
Исправил замечания ревьюера

- Добавил обработку отбрасывания необрабатываемого исключения в GlobalExceptionHandler
- Сделал корректную передачу DTO сущностей на контроллеры из сервисов
- Добавил аннотацию @UtilityClass в мапперы для ограничения возможности создания их экземпляров
- Оптимизировал метод ItemRepositoryImpl.getSearchingItemsByText

MaximusDenisov-YP committed:
Исправил замечания Checkstyle

avfyodorov reviewed (commented):
Добрый день, Максим!

Замечаний нет.
Работа принята.
